### ISSUE TITLE
`populate` on `afterFlush` resets uow's changeSet

### ISSUE THREAD

**Issue by abelland** [mikro-orm/mikro-orm#3345]
**Describe the bug**
on `afterFlush` method, uow's changeSet is emptied if we call `populate`

**To Reproduce**
```
import type { EventSubscriber, FlushEventArgs } from '@mikro-orm/core';
import { Collection, Entity, ManyToOne, MikroORM, OneToMany, PrimaryKey, Property, Subscriber } from '@mikro-orm/core';
import type { SqliteDriver } from '@mikro-orm/sqlite';

@Entity({ tableName: 'customers' })
class Customer {

  @PrimaryKey()
  id!: number;

  @Property()
  name: string = 'Foo';

  @OneToMany(() => Order, order => order.customer)
  orders = new Collection<Order>(this);

}

@Entity({ tableName: 'orders' })
class Order {

  @PrimaryKey()
  id!: number;

  @Property()
  value: number = 200;

  @ManyToOne(() => Customer)
  customer!: Customer;

}

@Subscriber()
class OrdersSubscriber implements EventSubscriber<Order> {

  async afterFlush(args: FlushEventArgs): Promise<void> {
    const changeSets = args.uow.getChangeSets();

    for (const changeSet of changeSets) {
      if (changeSet.entity instanceof Order) {
        await args.em.populate(changeSet.entity, ['customer']);
        console.log(args.uow.getChangeSets()) // <-- Empty Array
      }
    }
  }

}
```

**Expected behavior**
in mikro orm v4 change set is never altered, i am able to access it even after `populate` method. 

**Additional context**
i have an abstract class that allows to not having all afterFlush methods in a single subscriber (that we'll be unconfortable)

```
import {
  AnyEntity,
  ChangeSetType,
  EntityData,
  EntityName,
  EventSubscriber,
  FlushEventArgs
} from "@mikro-orm/core";
import _ from "lodash";

export abstract class BaseEventSubscriber<T extends AnyEntity>
  implements EventSubscriber<T>
{
  abstract onCreate(entity: AnyEntity<T>): Promise<void>;
  abstract onUpdate(
    entity: AnyEntity<T>,
    changedFields: EntityData<AnyEntity<T>>
  ): Promise<void>;

  getSubscribedEntities(): EntityName<T>[] {
    return [];
  }

  async afterFlush(args: FlushEventArgs): Promise<void> {
    // contains all the entity passed to .flush() method
    const changeSet = args.uow.getChangeSets();

    const entitiesNames = this.getSubscribedEntitiesNames();
    const entities = _.filter(changeSet, c =>
      _.includes(entitiesNames, c.name)
    );

    for (const entityElement of entities) {
      entityElement.type === ChangeSetType.CREATE &&
        (await this.onCreate(entityElement.entity));

      entityElement.type === ChangeSetType.UPDATE &&
        (await this.onUpdate(entityElement.entity, entityElement.payload));
    }
  }

  private getSubscribedEntitiesNames(): string[] {
    const subscribedEntity = this.getSubscribedEntities();
    return _.reduce(
      subscribedEntity,
      (acc, el) => {
        switch (typeof el) {
          case "string":
            return _.concat(acc, el);
          case "object":
            return _.concat(acc, el.name);
          case "function":
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            return _.concat(acc, el.name);
          default:
            return [];
        }
      },
      []
    );
  }
}

```

basically, after flush method is called for each concrete class that extends this one. if in one of them a `populate` is called, i am not able to access the change set of the unit of work anymore.

**Versions**

| Dependency | Version |
| - | - |
| node | 14 |
| typescript | 4.7.4 |
| mikro-orm | 5.2.3 |
| your-driver | postgresql |

**Comment by B4nan** [comment]
Not reproducible, and tbh this seems also literally impossible, calling `em.populate()` will never remove anything from a changeset, neither it will clear the UoW state.

My guess is a missing await somewhere. Let's reopen with a complete and failing reproduction, here is mine based on your code, and it works just fine:

```ts
import type { EventSubscriber, FlushEventArgs } from '@mikro-orm/core';
import { Collection, Entity, ManyToOne, MikroORM, OneToMany, PrimaryKey, Property, Subscriber } from '@mikro-orm/core';
import type { SqliteDriver } from '@mikro-orm/sqlite';

@Entity({ tableName: 'customers' })
class Customer {

  @PrimaryKey()
  id!: number;

  @Property()
  name: string = 'Foo';

  // eslint-disable-next-line @typescript-eslint/no-use-before-define
  @OneToMany(() => Order, order => order.customer)
  orders = new Collection<Order>(this);

}

@Entity({ tableName: 'orders' })
class Order {

  @PrimaryKey()
  id!: number;

  @Property()
  value: number = 200;

  @ManyToOne(() => Customer)
  customer!: Customer;

}

@Subscriber()
class OrdersSubscriber implements EventSubscriber<Order> {

  async afterFlush(args: FlushEventArgs): Promise<void> {
    const changeSets = args.uow.getChangeSets();

    for (const changeSet of changeSets) {
      if (changeSet.entity instanceof Order) {
        await args.em.populate(changeSet.entity, ['customer']);
        console.log(args.uow.getChangeSets()); // <-- Empty Array
      }
    }
  }

}

describe('GH issue 1003', () => {

  let orm: MikroORM<SqliteDriver>;

  beforeAll(async () => {
    orm = await MikroORM.init({
      entities: [Customer],
      dbName: ':memory:',
      type: 'sqlite',
    });
    await orm.getSchemaGenerator().createSchema();
  });

  afterAll(async () => {
    await orm.close(true);
  });

  test(`GH issue 1003`, async () => {
    const parent = orm.em.create(Customer, { name: 'asd', orders: [{ value: 123 }, { value: 456 }] });
    await orm.em.persist(parent).flush();

    parent.orders[0].value = 666;
    await orm.em.flush();
  });

});
```

**Comment by abelland** [comment]
> edit2: I havent really understood that code in your Additional context section. you are calling onCreate and onUpdate hooks manually, from insdie the afterFlush handler? that sounds quite wrong to me, those hooks need to be fired inside the flush context, not after it ends.

some actions need to be executed after flush, making sure data is saved in DB. i simply distinguish between create and update changeSetType.

**Comment by B4nan** [comment]
Actually, its enough to reload the entity and try again, then it reproduces. Weird, lets reopen.

```ts
const parent = orm.em.create(Customer, { name: 'asd', orders: [{ value: 123 }, { value: 456 }] });
await orm.em.persist(parent).flush();

parent.orders[0].value = 666;
await orm.em.flush();

orm.em.clear();
const o = await orm.em.findOneOrFail(Order, parent.orders[0]);
o.value = 3223232;
await orm.em.flush();
```

**Comment by B4nan** [comment]
Yes, this is indeed coming from the auto flush getting triggered via the `populate` call. 

```
    postCommitCleanup Error
        at UnitOfWork.postCommitCleanup (/Users/adamek/htdocs/mikro-orm/mikro-orm/packages/core/src/unit-of-work/UnitOfWork.ts:616:38)
        at UnitOfWork.commit (/Users/adamek/htdocs/mikro-orm/mikro-orm/packages/core/src/unit-of-work/UnitOfWork.ts:306:12)
        at async SqlEntityManager.flush (/Users/adamek/htdocs/mikro-orm/mikro-orm/packages/core/src/EntityManager.ts:814:5)
        at async SqlEntityManager.tryFlush (/Users/adamek/htdocs/mikro-orm/mikro-orm/packages/core/src/EntityManager.ts:831:7)
        at async SqlEntityManager.find (/Users/adamek/htdocs/mikro-orm/mikro-orm/packages/core/src/EntityManager.ts:107:5)
        at async EntityLoader.populateMany (/Users/adamek/htdocs/mikro-orm/mikro-orm/packages/core/src/entity/EntityLoader.ts:195:18)
        at async EntityLoader.populateField (/Users/adamek/htdocs/mikro-orm/mikro-orm/packages/core/src/entity/EntityLoader.ts:285:19)
        at async EntityLoader.populate (/Users/adamek/htdocs/mikro-orm/mikro-orm/packages/core/src/entity/EntityLoader.ts:64:7)
        at async SqlEntityManager.populate (/Users/adamek/htdocs/mikro-orm/mikro-orm/packages/core/src/EntityManager.ts:875:5)
        at async OrdersSubscriber.afterFlush (/Users/adamek/htdocs/mikro-orm/mikro-orm/tests/issues/GH1003.test.ts:43:9)
```

I am not sure if this is something we can fix internally (will try for sure before closing again), but you can definitely use the commit flush mode to get around it. If you want to disable it just for this particular EM you can fork it first (or disable it globally via ORM config):

```ts
const fork = orm.em.fork({ flushMode: FlushMode.COMMIT });
const o = await fork.findOneOrFail(Order, parent.orders[0]);
o.value = 3223232;
await fork.flush();
```

**Comment by B4nan** [comment]
Oh actually we explicitly disable auto flush when inside hooks, and apparently the flag is no longer set for the `afterFlush` hook, so this should be rather easy!

**Comment by B4nan** [comment]
Fixed via e3f34aa
